### PR TITLE
lua: add lua plugin

### PR DESCRIPTION
**Pull request by manuelm**
things before merging:
- [x] review work
- [x] because of shared error numbers, either crypto or lua plugin needs a rebase

### CODE REVIEW

File: src/plugins/lua/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+The plugin uses Lua to do magic things.
+
+## USAGE
+
+    kdb mount file.ini /lua lua script=/path/to/lua/some_lua_script.lua
+
+## Plugin Config ##
+
+No optional configuration values/flags yet
```
Comment (markus2330) [inline src/plugins/lua/README.md]: Should be passed to lua plugin

File: src/plugins/lua/lua.cpp [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+{
+	lua_State *L;
+	int printError;
+	int shutdown;
+} moduleData;
+
+static void *Lua_alloc(void *ud ELEKTRA_UNUSED, void *ptr,
+		size_t osize ELEKTRA_UNUSED, size_t nsize)
+{
+	if (nsize == 0) {
```
Comment (markus2330) [inline src/plugins/lua/lua.cpp]: newline before {

File: src/plugins/lua/lua.cpp [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+int elektraLuaOpen(ckdb::Plugin *handle, ckdb::Key *errorKey)
+{
+	KeySet *config = elektraPluginGetConfig(handle);
+	Key *script = ksLookupByName(config, "/script", 0);
+	if (script == NULL)
+		return 0; // success if no script to execute
+
+	if (keyString(script) == NULL)
```
Comment (markus2330) [inline src/plugins/lua/lua.cpp]: thats impossible, keyString never returns NULL.

**Comment by markus2330** [comment]
looks very nice!

File: src/plugins/lua/README.md (hunk 1 above)
Comment (manuelm) [inline src/plugins/lua/README.md]: what?

File: src/plugins/lua/lua.cpp (hunk 3 above)
Comment (manuelm) [inline src/plugins/lua/lua.cpp]: https://github.com/ElektraInitiative/libelektra/pull/286#discussion_r39558558 :smile_cat:

**Comment by markus2330** [comment]
- [ ] open 5

**Comment by manuelm** [comment]
jenkins build bindings please

**Comment by manuelm** [comment]
@markus2330 need some clarification on your comments please

File: src/plugins/lua/lua.cpp (hunk 3 above)
Comment (markus2330) [inline src/plugins/lua/lua.cpp]: That was pseudocode ;)

The correct way is to check for "/module", if this is present return 0. (It is guaranteed to be present during mounting and if mounted in system/elektra/modules). If its not a module, *then*, a not available "/script" is an error. It should be fixed for the python plugin, too. See src/plugins/jni/jni.c

Comment (manuelm) [inline src/plugins/lua/lua.cpp]: so `/module` exists during loading in order to check if the library is actually usable?

Comment (markus2330) [inline src/plugins/lua/lua.cpp]: Yes, /module is a special marker saying: don't expect any configuration. The plugin will be below system/elektra/modules and should return some standard contract. Its explained a bit in plugin interface, see elektraDocOpen.

Your code looks good now.

**Comment by manuelm** [comment]
@markus2330 In general ready for merging.

It's ok if you prefer to merge crypto plugin before and let me renumber the error numbers.

**Comment by markus2330** [comment]
Could you please rebase/fix error numbers

**Comment by manuelm** [comment]
jenkins build bindings please

**Comment by manuelm** [comment]
rebased and ready to go

File: src/plugins/lua/README.md (hunk 1 above)
Comment (markus2330) [inline src/plugins/lua/README.md]: The plugin configuration should be passed to the lua plugin.

**Pull request merged**